### ISSUE TITLE
Decoders

### ISSUE THREAD

**Issue by skaae** [Lasagne/Lasagne#391]
I would like to have layers that allows you to build an encoder/decoder structures in Lasagne. 

These layers are very similar to regular RNNs but with a few differences depending on the variant.
Heres a list of two common decoders:

1. A RNN that is initialized with the last hidden state of the decoder. The input is either None or the decoders prediction from the previous time step. 
2. A RNN with attention. The RNN decoder takes the entire encoded sequence as input and selectively focus on different parts of the encoded sequence. The input at each timestep is what you attended to and maybe also the prediction from the previous time step. 

For a GRU decoder both are easy to implement but for the LSTM layers its is complicated by both having cell and hidden states form the decoder. (Not really complicated we just need to settle on how to handle this see https://github.com/Lasagne/Lasagne/issues/337).

To be able to support feeding in the previous prediction to the next timestep i think the decoders would need to take an output network as input?

This is also related to https://github.com/Lasagne/Lasagne/issues/362. If we implement decoding layers there is no need for the RepeatLayer.

**Comment by f0k** [comment]
> A RNN that is initialized with the last hidden state of the decoder.

That's possible already, isn't it?

> The input is either None or the decoders prediction from the previous time step. [...] To be able to support feeding in the previous prediction to the next timestep i think the decoders would need to take an output network as input?

Can't you just stack the decoder on top of the encoder? But maybe I don't understand how it's supposed to work. Can you point us to some literature?

**Comment by skaae** [comment]
Yes you can sort of do it except that you need to give the decoder some input, which can ofcourse be 0's but that is kind of akward. 

If you want to use the prevous prediction as input then you need to have the output network inside the scan function. You need to process the first "column" of the decoder before you can process the next etc. Right now we rely on being able to process all positions in a recurrent layer before we start processing the next layer. This breaks if we need to use the previous output as input. 

Does that make sense? Otherwise i'll create a drawning :)

**Comment by f0k** [comment]
> Right now we rely on being able to process all positions in a recurrent layer before we start processing the next layer. This breaks if we need to use the previous output as input.

So the decoder is a stack of recurrent layers, but the first layer in this stack needs the output of the last layer in this stack (for the previous time step)?

**Comment by skaae** [comment]
Yes, except that the decoder does not need to be a stack for us to have a problem. As long as we have the dependency to the previous timestep we cannot do it with the current layers. 

I have a attached an image. If you look at the decoder you can see that the output from the decoder (U_i) is used as input in the next timestep (z_{i+1}) of the decoder. 

![Decoder](http://devblogs.nvidia.com/parallelforall/wp-content/uploads/sites/3/2015/06/Figure2_NMT_system.png)

**Comment by f0k** [comment]
> Yes, except that the decoder does not need to be a stack for us to have a problem. As long as we have the dependency to the previous timestep we cannot do it with the current layers.

But that's what you always have in a recurrent layer; the next timestep depends on the previous one. The only trouble with the architecture in the Figure is how to split it up into a input-to-hidden and hidden-to-hidden network. As the hidden state is defined as the data that is propagated from the previous timestep to the next, the hidden state here would be (z_i, u_i). Thus, the input-to-hidden network would just do something with the final state of the encoder that can only modify z_i, and the hidden-to-hidden network would do something with (z_i, u_i) to provide (z_i+1, u_i+1). That should all be possible with the current code. It might still be nicer to provide a simpler way than fiddling with CustomRecurrentLayer, of course, and I guess it wouldn't work for the current LSTM and GRU implementations anyway.
So, yes, a decoder layer with a custom output network might make things easier to use. Would that help anything for LSTM and GRU, though?

**Comment by skaae** [comment]
I'm not sure i understand your explanation completely, so we might agree.  

The problem is the `hidden_to_output` network that we need to put inside scan. At least for GRU and LSTM that is currently not possible. I don't know if you can do it with the custom RNN layer.

In math we want: (I use subscript t to denote time step)

h_t = f(x_t, h_{t-1}, **o_{t-1}**) 
o_t = g(h_t)  

h_t : hidden state or RNN
o_t : output. 
g    : some output network, e.g. a softmax layer 
f     : some recurrent layer function e.g. LSTM, GRU etc

Maybe the figure was not that informative.

**Comment by f0k** [comment]
I was saying that the hidden-to-output network could just be part of the hidden-to-hidden network if you want to implement it with what we currently have.

(h_t, o_t) = y(x_t, (h_{t-1}, o_{t-1}))

(h_t, o_t) : hidden state of RNN
y : the hidden-to-hidden network

The hidden-to-hidden network would have a ConcatLayer in the end that merges h_t and o_t. So referring to your definitions above, it would be of the form:
h_t = f(x_t, h_{t-1}, o_{t-1})
y = concat(h_t, g(h_t))

You'd then need a SliceLayer on top of the decoder that separates out o_t again. (Unless we solve #337, then we don't need to merge them in the first place.) Does that make sense?

I still agree that this merging and slicing could be a plausible reason to add a hidden-to-output network. Would it be possible to extend `CustomRecurrentLayer` with that? The user-provided hidden-to-hidden network would need to have two InputLayers in this case. Thinking some more, couldn't `CustomRecurrentLayer` accept a single network of two inputs instead of two single-input networks already, wouldn't that make it more flexible?

**Comment by skaae** [comment]
I have never really used CustomRecurrentLayer but I think it would be possible to have single network with two inputs. I don't know if i prefer that option because I think it complicates the setup?
In any case I'm more interested in the GRU and LSTM layers.

One option is to try to generaize the GRU and LSTM layers to accept a hidden_to_output layer. 

Another option is to create a "feedforward" LSTM layer and a "feedforward" GRU layer, which should bascially inplement the stepfunction from the current LSTM layers. We could then let the user specify a list of "feedforward" layers that we could stack inside a scan. Dunno if that is possible but it would allow us to support multilayer decoders rather easily.

**Comment by craffel** [comment]
> One option is to try to generaize the GRU and LSTM layers to accept a hidden_to_output layer.

See https://github.com/Lasagne/Lasagne/issues/357  One nice thing about only allowing a custom hidden->output connection is that we can still precompute the input dot products.

> Another option is to create a "feedforward" LSTM layer and a "feedforward" GRU layer, which should bascially inplement the stepfunction from the current LSTM layers. We could then let the user specify a list of "feedforward" layers that we could stack inside a scan. Dunno if that is possible but it would allow us to support multilayer decoders rather easily.

This is an interesting idea.

**Comment by skaae** [comment]
Den 21/08/2015 19.39 skrev "Colin Raffel" <notifications@github.com>:
>>
>> One option is to try to generaize the GRU and LSTM layers to accept a
hidden_to_output layer.
>
> See #357 One nice thing about only allowing a custom hidden->output
connection is that we can still precompute the input dot products.
>>
>> Another option is to create a "feedforward" LSTM layer and a
"feedforward" GRU layer, which should bascially inplement the stepfunction
from the current LSTM layers. We could then let the user specify a list of
"feedforward" layers that we could stack inside a scan. Dunno if that is
possible but it would allow us to support multilayer decoders rather easily.
>
> This is an interesting idea.
>
>
I'll try to write an example some time next week.  Maybe it'll also #357.

**Comment by braingineer** [comment]
Hello!

I came across this while trying to use Lasagne to generate a language model.  I have looked at the [lstm text generation example](https://github.com/Lasagne/Recipes/blob/master/examples/lstm_text_generation.py) in recipes, but I think this discussion is more in line with what I want.  That example (and the recurrent.py example) use the last output from RNN for BPTT.  However, I want to do sequence-element-wise backprop.  

To visually illustrate, from Andrej Karpathy's blog:

![an rnn illustration](http://karpathy.github.io/assets/rnn/diags.jpeg) 

What is currently accomplished with SliceLayer(an_rnn_layer, -1, 1) as the avenue for the cost function, is the many to 1.  The 1 is where the loss is coming from.

What I would like to do is the many to many (the encoder-decoder approach, as has been discussed here). 

Can I interpret the earlier discussion as indicating one needs to modify the scan routine in order to do this?  

One hack-y thought I had was to use a slice layer at every sequence step, then have a shared weight dense layer on top of all of those slice layers.  Then, I could backprop the loss easily. 

Thanks for your time!
Brian

**Comment by f0k** [comment]
> What I would like to do is the many to many (the encoder-decoder approach, as has been discussed here).

Just to be clear, you mean the second to last example in the illustration, right? (And not the last one, which is also labeled "many to many", but is not an encoder-decoder architecture?)

> Can I interpret the earlier discussion as indicating one needs to modify the scan routine in order to do this?

No, I think you'd define the encoder (which would look like the "many-to-one" example, with a SliceLayer(..., -1) in the end) and get its output expression via `lasagne.layers.get_output()`, then define the decoder (which would look like the "one-to-many" example) and set its `hid_init` to the output expression you captured before. For training, you will need to collect the parameters of the encoder and decoder with two separate `lasagne.layers.get_all_params(trainable=True)` calls.
We might refactor the code such that it's possible to directly pass a network for `hid_init`, that would make it a tad easier.
Disclaimer: I've just started with RNNs and didn't need an encoder-decoder architecture yet, so this is just what I think it will look like. But I'm not sure the "one-to-many" decoder is possible with Lasagne right now, maybe the only way is a "many-to-many" decoder like the very last example that has its input repeated.

**Comment by braingineer** [comment]
re: last illustration.

Yea!  Sorry if that was confusing.  I was thinking of the encoder-decoder approaches as embedding into a state space in the input, then retrieving results from that state on the output.  The many-many approach though, is used by Karpathy in his latest cvpr paper, [found here in section 3](http://cs.stanford.edu/people/karpathy/cvpr2015.pdf).  

re: interpretation discussion. 

Yep!  Though, the one-to-many should be easy with Lasagne, I think, because it's just the initial hidden layer.  The many-to-one is easy, as you said, with the slice layer.  In peaking around at other implementations, it seems most people do this because it is easier.  I haven't found many yet that utilize the error at every time step in the BPTT.

In the paper I linked, Karpathy uses the output at each time step as the embedde d vector for that input word.  He uses that embedded vector along with an image vector and minimizes a structured margin loss function using cosine similarity of the word vector and the image vector(s).  

But, that's not the only error you could have for each time step.  You could also image that at each time step, you have a final layer of sigmoid activations treated as P(next word|words so far), which is more like a traditional chain-emission model (linear crf, hmm, etc).  

Both of the things I described require backpropping the error back at each time step.  I'm also just beginning with RNNs and Lasagne, so I'm not sure if this is possible with how Lasagne is set up.

**Comment by f0k** [comment]
> Both of the things I described require backpropping the error back at each time step. I'm also just beginning with RNNs and Lasagne, so I'm not sure if this is possible with how Lasagne is set up.

Both of this is possible, also in combination, with some appropriate reshaping in between.

I'm still not sure about the "one-to-many" architecture, though. Setting the initial hidden state to the output of some network is possible with an additional `get_output()` call as I described, but I think currently there is no way to have a recurrent layer that does *not* take an input at each time step.

**Comment by braingineer** [comment]
Hi f0k,

Sorry for the delayed response.  Things get busy =).

Thank you for your response!  

>>Both of thi si spossible, also in combination, with some appropriate reshaping in between.

I think my confusion comes from not seeing how you would construct a loss function that would take into account every sequence step.  In your recurrent example, you use a slice layer and then a fully connected layer to a single node.  But, this may not be the appropriate chain anymore for this.   

**Actually**, I have just noticed that the example on the recurrent page now explains what you've said.  I'm not sure if it was there before (if it was, apologies for the confusions!), but yay and thanks.

re: decoder. 
My intuition on creating a decoder was the have the hid_init be the output of the encoder and sequences always start with a start symbol.  The training would be the same as the other network I've been referencing would be (where each time step's output is treated independently).

**Comment by MInner** [comment]
> I'm still not sure about the "one-to-many" architecture, though. Setting the initial hidden state to the output of some network is possible with an additional get_output() call as I described, but I think currently there is no way to have a recurrent layer that does not take an input at each time step.

Just asking for an update from those who are aware of current state of matter: is it correct that the only way to handle this now is to use the same hack @f0k suggested to @skaae : use a `lasagne.layers.CustomRecurrentLayer` with a special `input-to-hidden` and `input-to-hidden` Layers that would Merge both "real output" and "real hidden state" into a single "virtual hidden value" and unpack (Slice) them "from other side" (on the next iteration)? And that also limits us with only `lasagne.layers.RecurrentLayer`-like recurrency (no space for `LSTMLayer` or `GRULayer`), right?

(sidenote: I remember well theano being not so happy with me not using all the variables in the computational graph; wouldn't the same issue arise if input value of the `CustomRecurrentLayer` layer is techincally not used in the decoder? Or I can just add it, like `+input`-term and always pass zeros there, though.)

**Comment by f0k** [comment]
> Just asking for an update from those who are aware of current state of matter: is it correct that the only way to handle this now is [...]

Yes, nothing has changed in this respect. There's a pretty active discussion on refactoring the recurrent layers, though: #425. You're very welcome to chime in!
